enhancements/sig-policy/98-long-term-compliance-history/README.md

```diff
@@ -0,0 +1,455 @@
+event would be recorded in etcd with the key format of
+`compliance-events/<timestamp from Kubernetes event>-<event-uid>`. For example,
+`compliance-events/2023-07-25T13:15:08Z-4b400c5f-3828-4138-ad4f-03de4d69655b`. The value would be in the format of the
+`POST` request for the `/api/v1/compliance-events` HTTP endpoint on the hub.
+
+The Governance Policy Framework addon pod would run a separate goroutine that would periodically poll etcd for new
+compliance events and try to record them on the hub. It would do this in order by sorting by the timestamps in the keys.
+When successful, it would delete the entry in etcd.
+
+By default, the embedded etcd would use an [emptyDir](https://kubernetes.io/docs/concepts/storage/volumes/#emptydir)
+volume which requires no persistent volume configuration on the managed cluster and would survive container restarts. If
+the pod is removed from the node, the data is lost however. For this reason, OCM could be configured to request a
+persistent volume for better accuracy. This would be configured using an
+[AddOnDeploymentConfig](https://github.com/open-cluster-management-io/enhancements/blob/main/enhancements/sig-architecture/58-addon-configuration/README.md)
+`customizedVariables` entry. The variable would be `{"name": "UsePVForComplianceEventCache", "value": true}`. Having
+this variable set would trigger the Governance Policy Addon controller to create a `PersistentVolumeClaim` instead of
+using `emptyDir`. We should also consider an "out of the box" policy that checks for instances where a persistent volume
+is not used to warn the user of this.
+
+##### Managed Cluster Phase 3 - Additional Metadata
+
+In my opinion, using Kubernetes events as a message queueing system is abusing what it is meant to be, which is
+information for the cluster user. This generates a lot of noise that is not relevant to the user. Additionally, the new
+[Event](https://kubernetes.io/docs/reference/kubernetes-api/cluster-resources/event-v1/#Event) API only allows a `note`
+of 1kB. This is very limiting for when the policy components would need to migrate to this new `Event` API. That being
+said, the old `Event` API is not expected to be removed in the near future though.
+
+It is proposed to expose the embedded etcd GRPC endpoint and have policy controllers directly record compliance events
+using it instead of recording Kubernetes events when the embedded etcd is backed by a persistent volume. The GRPC
+endpoint would be protected with authentication and the credentials would be stored in a secret called
+`governance-policy-framework-cache` in the managed cluster namespace. Each policy controller would read this secret on
+start up. Recording the compliance event directly from the controller would allow recording rich metadata about the
+compliance event. One example is the diff between what is in the policy and what is on the cluster, similar to what
+ArgoCD does.
```

Comment:
Is a shared secret the best way to do this? I was expecting the same kind of authorization that is protecting the hub API endpoint (requiring "write access to the status subresource of the policies.policy.open-cluster-management.io resource").

The shared secret does make the embedded etcd simpler, since it doesn't need a proxy in front of it or anything...

Reply:
Since the embedded etcd is an internal implementation detail, I'd like to avoid using the Kubernetes API for authorization since this adds an unnecessary additional load on the Kubernetes API and would slow recording the compliance events.